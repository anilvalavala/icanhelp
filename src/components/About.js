import React from 'react';

const About = () => {
    return(
        <div>
            <h1>What is this about?</h1>
            <br />
            <p>
                Yes, today the world is full of sophisticated computers, machines, devices, food, what not, everything. <br />
                But still there are a section of people who are looking for help. <br /><br />
                There are many people who want to extend their helping hand to the society <br />
                But there is no bridge connecting such kind hearted people to the needy ones. <br /> <br />

                So, this application is intended for those people who want to donate something. <br />
                One can simply update their donation details and everyone on web can see it,  
                especially NGOs which are working for upliftment of less privileged. <br />

                This is just a basic verion of my idea. 
                Will keep updating this app to make it more user friendly.
            </p>
        </div>
    );
}

export default About;