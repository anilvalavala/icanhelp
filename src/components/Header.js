import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => {
    return(
    <div className="header">
        <Link className="header__link" to="/">Home</Link>
        <Link className="header__link" to="/about">About</Link>
        <Link className="header__link" to="/feedback">Feedback</Link>
    </div>
    );
};

export default Header;