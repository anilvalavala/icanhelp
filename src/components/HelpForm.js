import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import uuid from 'uuid';

export default class HelpForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            id: props.helpItem ? props.helpItem.id : uuid(),
            title: props.helpItem ? props.helpItem.title : '',
            description: props.helpItem ? props.helpItem.description : '',
            fromDate: props.helpItem ? moment(props.helpItem.fromDate) : moment(),
            fromDateFocused: true,
            toDate: props.helpItem ? moment(props.helpItem.toDate) : moment(),
            toDateFocused: true,
            phone: props.helpItem ? props.helpItem.phone : '',
            email: props.helpItem ? props.helpItem.email : '', 
            isPageInvalid: false,
            titleError: '',
            descriptionError: '',
            phoneError: '',
            emailError: '',
            dateRangeError: ''
        }
    }
    
    setTitle = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    }
    setDescription = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    setEmail = (e) => {
        const email = e.target.value;
        if(email && email.match(/^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/))
        {
            this.setState(() => ({ email }));
        }
    }
    setPhoneNumber = (e) => {
        const phone = e.target.value;
        if(phone && phone.match(/^[6-9]\d{9}$/))
        {
            this.setState(() => ({ phone }));
        }
    }
    setFromDate = (fromDate) => {
        if(fromDate){
            this.setState(() => ({ fromDate }));
        }
    }
    setFromDateFocus = ({ focused }) => {
        this.setState(() => ({ fromDateFocused: focused }));
    }
    setToDate = (toDate) => {
        if(toDate){
            this.setState(() => ({ toDate }));
        }
    }
    setToDateFocus = ({ focused }) => {
        this.setState(() => ({ toDateFocused: focused }));
    }
    onSubmit = (e) => {
        e.preventDefault();
        //Title validation
        if(!this.state.title){
            this.setState(() => ({
                titleError: 'Title field can\'t be empty',
                isPageInvalid: true
            }));
        }else {
            this.setState(() => ({
                titleError: '',
                isPageInvalid: false
            }));
        }
        //Description validation
        if(!this.state.description){
            this.setState(() => ({
                descriptionError: 'Description field can\'t be empty',
                isPageInvalid: true
            }));
        }else {
            this.setState(() => ({
                descriptionError: '',
                isPageInvalid: false
            }));
        }
        //Phone number vaidation
        if(!this.state.phone){
            this.setState(() => ({
                phoneError: 'Phone number must start with 6/7/9 and should contain 10 digits',
                isPageInvalid: true
            }));
        }else {
            this.setState(() => ({
                phoneError: '',
                isPageInvalid: false
            }));
        }
        //Email validation
        if(!this.state.email){
            this.setState(() => ({
                emailError: 'Email id is not valid',
                isPageInvalid: true
            }));
        }else {
            this.setState(() => ({
                emailError: '',
                isPageInvalid: false
            }));
        }

        if(!this.state.isPageInvalid)
        {
            this.props.onSubmit({
                id: this.state.id,
                title: this.state.title,
                description: this.state.description,
                fromDate: this.state.fromDate.valueOf(),
                toDate: this.state.toDate.valueOf(),
                phone: this.state.phone,
                email: this.state.email,
        });
        }
    }
    render(){
        return(
            <div>
            <form onSubmit={this.onSubmit}>
                Help Form to Add/Edit Help item <br />
                <div>
                    Title: 
                    <input type="text" value={this.state.title} onChange={this.setTitle}/> 
                    {this.state.titleError && <p>{this.state.titleError}</p>}
                </div> <br />
                <div>
                    Description: 
                    <textarea value={this.state.description} onChange={this.setDescription}></textarea> 
                    {this.state.titleError && <p>{this.state.titleError}</p>}
                </div> <br />
                <div>
                    Available From: 
                    <SingleDatePicker
                        date={this.state.fromDate}
                        onDateChange={this.setFromDate}
                        focused={this.state.fromDateFocused}
                        onFocusChange={this.setFromDateFocus}
                        numberOfMonths={1}
                    />
                </div> <br />
                <div>
                    Available Upto:
                    <SingleDatePicker
                        date={this.state.toDate}
                        onDateChange={this.setToDate}
                        focused={this.state.toDateFocused}
                        onFocusChange={this.setToDateFocus}
                        numberOfMonths={1}
                    />
                </div> <br />
                <div>
                    Email: <input type="text" value={this.state.email} onChange={this.setEmail}/>
                </div> <br />
                <div>
                    Phone: <input type="text" value={this.state.phone} onChange={this.setPhoneNumber}/> <br />
                </div> <br />
                <button>Save</button>
            </form>
            </div>
        );
    }
}