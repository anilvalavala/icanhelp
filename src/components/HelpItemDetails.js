import React from 'react';
import { Link } from 'react-router-dom';

const HelpItemDetails = (props) => {
    return(
        <div>
            <div>
                <Link to={`/editHelp/${props.helpItem.id}`}> {props.helpItem.title} </Link> <br />
                Description: {props.helpItem.description} <br />
                Available From: {props.helpItem.fromDate} <br />
                Available Upto: {props.helpItem.toDate} <br />
                Email: {props.helpItem.email} <br />
                Phone: {props.helpItem.phone} <br />
                <button onClick={
                    () => {
                        props.dispatch(deleteHelpItem(props.helpItem.id));
                    }
                }>Delete</button>
            </div>
        </div>
    );
}; 

export default HelpItemDetails;