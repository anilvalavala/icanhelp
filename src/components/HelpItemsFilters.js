import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';
import { filterByText, sortByDate, startDateFilter, endDateFilter } from '../actions/Filters';

export class HelpItemsFilters extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            startDate: props.filters.startDate,
            endDate: props.filters.endDate,
            calendarFocused: null
        }
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.setState(() => ({
            startDate,
            endDate
        }));
        this.props.startDateFilter(startDate);
        this.props.endDateFilter(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onFilterByTextChange = (e) => {
        this.props.filterByText(e.target.value);
    };

    onSortByChange = (e) => {
        this.props.sortByDate(e.target.value);
    };

    render(){
        return(
            <div>
                Filter By: 
                    <input type="text" value={this.props.filters.filterBy} 
                       onChange={this.onFilterByTextChange} />
                Sort By: 
                    <select onChange={this.onSortByChange}>
                       <option value='fromDate'>From Date</option>
                       <option value='toDate'>To Date</option>
                    </select>
                    <br />
                Show items between:
                    <DateRangePicker
                        startDate={moment(this.state.startDate)}
                        startDateId="helpItemsStartDate"
                        endDate={moment(this.state.endDate)}
                        endDateId="helpItemsEndDate"
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        showClearDates={false}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        keepOpenOnDateSelect={false}
                    />
            </div>
        );
    }
}

const MapStateToProps = (state) => ({
    filters: state.filters
});

const MapDispatchToProps = (dispatch) => ({
    startDateFilter: (startDate) => { dispatch(startDateFilter(startDate)) },
    endDateFilter: (endDate) => { dispatch(endDateFilter(endDate)) },
    filterByText: (filterBy) => { dispatch(filterByText(filterBy)) },
    sortByDate: (sortBy) => { dispatch(sortByDate(sortBy)) }
});

export default connect(MapStateToProps, MapDispatchToProps)(HelpItemsFilters);