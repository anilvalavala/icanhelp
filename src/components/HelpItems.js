import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import HelpItemDetails from './HelpItemDetails';
import filteredHelpItems from '../selectors/HelpItems';
import { deleteHelpItem } from '../actions/HelpItems';

//we are exporting unconnected version also to allow testing this component
export const HelpItems = (props) => {
    return(
        <div>
        {
            props.helpItems.length === 0 ? (
                <p>No Help Items Found</p>
            ) : (
                <div>
                <h1>Help Items</h1>
                {
                    props.helpItems.map(
                        (helpItem) => (
                            <div key={helpItem.id}>
                                <HelpItemDetails 
                                    helpItem={helpItem}
                                />
                                <br />
                            </div>
                        )
                    )
                }
                </div>
            )
        }
        </div>
    );
};

const mapStateToProps = (state) => ({
    helpItems: filteredHelpItems(state.helpItems, state.filters)
});

export default connect(mapStateToProps)(HelpItems);