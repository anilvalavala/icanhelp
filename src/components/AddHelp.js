import React from 'react';
import { connect } from 'react-redux';
import HelpForm from './HelpForm';
import { startAddHelpItem } from '../actions/HelpItems';

export class AddHelp extends React.Component {
    onSubmit = (helpItem) => {
        this.props.startAddHelpItem(helpItem);
        this.props.history.push('/');
    };
    render(){
        return(
            <div>
                Add Help
                <HelpForm onSubmit={this.onSubmit} />
            </div>
        );
    };
}

// const AddHelp = (props) => {
//     return(
//         <div>
//         Add Help
//         <HelpForm onSubmit={(helpItem) => {
//             //The below line of code makes testing difficult
//             //Because addHelpItem is a dependency from another component
//             //props.dispatch(addHelpItem(helpItem));
//             props.onSubmit(helpItem)
//             props.history.push('/');
//         }}/>
//         </div>
//     );
// };

const mapDispatchToProps = (dispatch) => {
    return {
        startAddHelpItem: (helpItem) => dispatch(startAddHelpItem(helpItem))
    };
};


//When we do not have anything to connect from state to props but still we might need access to dispatch
//In this scenario we will use below syntax, connect without params
//export default connect()(AddHelp);

//connect has two params
//mapStateToProps - helps to attach state to props object, to consume in unconnected component
//mapDispatchToProps - helps to attach functions to props that can be used to dispatch actions
//mapDispatchToProps helps to abstract out dispatch calls from unconnected components
export default connect(undefined, mapDispatchToProps)(AddHelp);