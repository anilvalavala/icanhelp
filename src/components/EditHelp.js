import React from 'react';
import { connect } from 'react-redux';
import { editHelpItem } from '../actions/HelpItems';
import HelpForm from './HelpForm';

export class EditHelp extends React.Component {
    
    onSubmit = (helpItem) => {
        this.props.editHelpItem(helpItem);
        this.props.history.push('/');
    };
    render(){
        return(
            <div>
            <HelpForm 
                helpItem={this.props.helpItem}
                onSubmit={this.onSubmit} />
        </div>
        );
    };
}

// const EditHelp = (props) => {
//     return(
//         <div>
//             <HelpForm 
//                 helpItem={props.helpItem}
//                 onSubmit={(helpItem) => {
//                     props.dispatch(editHelpItem(helpItem));
//                     props.history.push('/');
//             }} />
//         </div>
//     );
// };

const mapStateToProps = (state, props) => {
    return({
        helpItem: state.helpItems.find((helpItem) => {
            return helpItem.id === props.match.params.id
        })
    })
};

const mapDispatchToProps = (dispatch) => {
    return {
        editHelpItem: (helpItem) => dispatch(editHelpItem(helpItem))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditHelp);