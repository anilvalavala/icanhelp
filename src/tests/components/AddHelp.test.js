import React from 'react';
import { shallow } from 'enzyme';
import { AddHelp } from '../../components/AddHelp';
import TestHelpItems from '../fixtures/TestHelpItems';

let addHelpItemSpy, historySpy, wrapper;

beforeEach(() => {
    addHelpItemSpy = jest.fn();
    historySpy = { push: jest.fn() };
    wrapper = shallow(<AddHelp addHelpItem={addHelpItemSpy} history={historySpy} />);
});

test('Test AddHelp component, if it is rendering correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Test AddHelp component, to check onSubmit', () => {
    wrapper.find('HelpForm').prop('onSubmit')(TestHelpItems[0]);
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(addHelpItemSpy).toHaveBeenLastCalledWith(TestHelpItems[0]);
});