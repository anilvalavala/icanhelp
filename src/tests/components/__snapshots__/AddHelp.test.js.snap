// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AddHelp component, if it is rendering correctly 1`] = `
<div>
  Add Help
  <HelpForm
    onSubmit={[Function]}
  />
</div>
`;
